### ISSUE TITLE
Wrap MPIX_Query_cuda_support

### ISSUE THREAD

**Issue by FreddieWitherden** [mpi4py/mpi4py#162]
Our application, PyFR, makes very successful use of mpi4py and has support for CUDA-aware MPI implementations.  Here, however, our biggest issue is knowing if the MPI distribution we are running under is CUDA aware or not.  Although there does not appear to be a perfect solution, for OpenMPI derivatives, the following is reasonable:

https://www.open-mpi.org/faq/?category=runcuda#mpi-cuda-aware-support

with the API of interest being MPIX_Query_cuda_support.  It would therefore be nice if mpi4py could expose this attribute (None/False/True).

**Comment by FreddieWitherden** [comment]
It also appears to be supported by MPICH: https://www.mpich.org/static/docs/v4.0a1/www3/MPIX_Query_cuda_support.html

**Comment by dalcinl** [comment]
@leofang and I pushed for Open MPI and MPICH to provide this API with uniform runtime semantics. @leofang has some preliminary code in his personal mpi4py fork, although it needs a few minor tweaks.

**Comment by dalcinl** [comment]
> It also appears to be supported by MPICH

MPICH has more than one API, depending on support for CUDA/ZE/HIP. Not sure if Open MPI or MVPICH support anything else beyond CUDA.

I had troubles in the past to make OMPI and MPICH teams agree on the semantics of the MPI standard itself, so I have no hope of making them agree on non-standard MPI APIs like querying GPU support. This also means that we would need to expose our own API, let say `Query_gpu_support()` to let say return `None` if not available or a string with the name of the backend (e.g.`"CUDA"` or `"HIP"`).

PS: I'm hesitant to add support for non standard stuff in mpi4py, because our team is very small, and user needs are infinite, and if things break and do not work as expected, the first place everyone makes pressure is downstream, and not upstream. However, I understand this is important enough, so I'm willing to break my own rules as long as we can define something that is future-proof. For example, could it happen in the future that some MPI may provide support for more than one backend? So how should we define the return value of the Python function `Query_gpu_support()`? Perhaps a list of strings with backend names, and empty list (or None?) if there is no GPU support at all?

**Comment by FreddieWitherden** [comment]
The MPI libraries are a complete and utter mess in this regard, and their APIs are usually an example of what not to do.  In an ideal world we've have something like `MPIX_Query_Param(const char* name, char *out, size_t *outsz)` which mpi4py could then wrap and leave everything else up to the user.  Sadly, we seem to be getting one new API function per platform (CUDA, HIP, Intel's level0, ...).

One option on the mpi4py level might be to see if it is possible to expose a ctypes handle to the MPI library which mpi4py is linked against.  This way the user could, at their peril, attempt to dynamically resolve the relevant symbols and call them.

If providing an API I think we probably want to account for the tri-state nature of the query: API is not provided, provided and returns True, provided and returns False.  Maybe something like a `gpu_support` object which has properties if a given API is supported and then that is set to either true or false.  So, `gpu_support.cuda` which would be either True, False, or an AttributeError.

**Comment by leofang** [comment]
My old PR that @dalcinl referred to is here: https://bitbucket.org/mpi4py/mpi4py/pull-requests/18. Feel free to work based on it and send a PR (to GitHub; we have abandoned Bitbucket) as I am not sure if I'd have time to polish it in the next few weeks.

Let me extend what @dalcinl said: The situation has become a lot messier since our initial work. Specifically, IIRC the semantics of `MPIX_Query_cuda_support` in Open MPI and MPICH are slightly different, with regard to the case of "the support is built but not enabled at runtime". (Thought I'd need some time to recollect the exact difference, sorry.) Also, MPICH has an umbrella API `MPIX_GPU_query_support` (which I think serves our purpose better) not seen in other MPI implementations.

Another nasty issue is either MPIs could be built with CUDA/HIP support disabled, but using UCX that has such support enabled (and, optionally, enabled/disabled at runtime). AFAIK there's no way to communicate the CUDA/HIP support between MPI and UCX, and there is no way either for us to know if the MPI is using UCX (so that we can go and ask if UCX supports any accelerator). 

In short: The whole query is a mess unless MPI maintainers are willing to sit down and work with us (@jsquyres @hzhou @harisubramoni), and it's such a mess that I do not have enough bandwidth for, at least not in the next few weeks.

**Comment by leofang** [comment]
By the way, forgot to reply:
> Maybe something like a `gpu_support` object which has properties if a given API is supported and then that is set to either true or false. So, `gpu_support.cuda` which would be either True, False, or an AttributeError.

It should just be a plain function returning something (bool, string, a list of bool or string, whatever). It does not seem pythonic to me to implement such a query utility as an object.

**Comment by dalcinl** [comment]
Using an `MPI_Info` with key-value pairs seems like the ideal interface for MPI implementations to have absolute freedom to advertise GPU support in whatever format and detail they want.

I'm wondering whether the new routine `MPI_Get_hw_resource_info` introduced in MPI-4.1 could somehow be used to advertise GPU support. Even if that standard routine is not appropriate, a new `MPIX_Query_support` or whatever name, and agreed upon MPICH and Open MPI, would do just fine. The output is a `MPI_Info` object, and it is up to the user how to parse the key-value pairs.

**Comment by FreddieWitherden** [comment]
Yes, as far as I am aware neither OpenMPI or MPICH provide a simple means (non-standard or otherwise) to determine if I can safely pass GPU pointers from vendor X to MPI functions and not get a segfault.  As such we still have to expose an option in a configuration files.

An `MPI_Info` based solution would be excellent if the vendors could adopt it.

**Comment by dalcinl** [comment]
@hzhou @jsquyres Would it be possible for the two major MPI implementations to agree on a let say `int MPIX_Query_features(MPI_Info*)` routine that would fill `MPI_Info` with implementatipn-defined key/value pairs to advertise support for features like GPU support?

Alternatively, and perhaps even better, what about exposing GPU support details via a set of implementation-defined key/value pairs in `MPI_INFO_ENV`?

**Comment by hzhou** [comment]
Moving forward, the standardized way of control and query GPU support is via `mpi_memory_alloc_kinds` info hints. Ref. https://github.com/mpi-forum/mpi-issues/issues/580

cc @raffenet

**Comment by dalcinl** [comment]
* `mpi_memory_alloc_kinds` is now standard in MPI 4.1.
* IIUC, MPICH 4.2 will support MPI 4.1 and add support for `mpi_memory_alloc_kinds`.
* Other MPI implementations could add support for it, even if they do not support full MPI 4.1, as these `MPI_Info`-based APIs are API and ABI agnostic. Users of these MPI implementations should advocate for the respective development teams to implement the needed support.

All that being said, I consider this issue resolved from the mpi4py side.
